### ISSUE TITLE
Multiple "random" looking like failures in colour-science unit tests since 3.4.0 release

### ISSUE THREAD

**Issue by KelSolaar** [pytest-dev/pytest-xdist#979]
Hello,

It took me a few days to pinpoint the [source](https://github.com/colour-science/colour/actions/runs/7028449640/job/19124427128) [of](https://github.com/colour-science/colour/actions/runs/7027933293) [our](https://github.com/colour-science/colour/actions/runs/7009452236) [broken](https://github.com/colour-science/colour/actions/runs/7001719291) builds and it comes down to the 3.4.0 release 3 weeks ago. It seems like this release (and 3.5.0) is now causing issues with our caches as if I disable them the [unit tests are passing](https://github.com/colour-science/colour/actions/runs/7015999647/job/19127292945).

Looking at the [diff between](https://github.com/pytest-dev/pytest-xdist/compare/v3.3.1...v3.4.0) 3.3.1 which is working and 3.4.0 which is not, the only change that I think could cause this is that one: https://github.com/pytest-dev/pytest-xdist/commit/230ba6ad1057574c9f3d42a97f890788cd9ec6c3

To confirm the finding, I have banned the 3.4.0 and 3.5.0 and our builds are now passing again: https://github.com/colour-science/colour/actions/runs/7029639824/job/19127625456

Keen to hear your thoughts on this one.

Cheers,

Thomas

**Comment by RonnyPfannschmidt** [comment]
At first glance this looks like a unrelated mismatch

The logic change doesn't Look like its touching any code related to floating point precision

Do the related libraries require per thread configuration of floating point precision

It's possible that different versions simply trigger a issue with execnet wrt running pytest on the main thread versus a different thread

**Comment by KelSolaar** [comment]
> The logic change doesn't Look like its touching any code related to floating point precision

Indeed!

> Do the related libraries require per thread configuration of floating point precision

I don't think they need but Numpy and Scipy have many multi-threaded routines so maybe this is causing some issues here.

I forked your repo and starting to walkup the commit chains and everything ran fine until this commit: https://github.com/pytest-dev/pytest-xdist/commit/3fe877bc6d7211f31e55d2d561803f76ea95b935. Seems like changing the ordering of the tests was enough to create the behaviour we are seeing, which is mildly concerning.

I checked `execnet` releases and the last one is from July, so it is not that. I'm not familiar with it at all but it seems like it is adopts a no-sharing model so the mystery deepens! Out of curiosity, is there a way to export a list of the tests a given worker is meant to process. I would like to see if I can repro the issue locally by simply making a run in the same order.

**Comment by RonnyPfannschmidt** [comment]
If you run with report log,the json lines output contains worker IDs

You can then use jq to extract the details

**Comment by KelSolaar** [comment]
> If you run with report log,the json lines output contains worker IDs
> 
> You can then use jq to extract the details

Thanks for the suggestion!

**Comment by KelSolaar** [comment]
I looked at a particular cache and some of its values:

Value cached on a [successful run](https://github.com/colour-science/colour/actions/runs/7044291964/job/19172696748#step:13:453):

```python
array([[2.500000000000000, 0.200000000000000, 2.000000000000000,
        4.000000000000000],
```

Value cached on an [unsuccessful run](https://github.com/colour-science/colour/actions/runs/7044291964/job/19172697042#step:13:7647):

```python
array([[2.500000000000000, 0.200000002980232, 2.000000000000000,
        4.000000000000000],
```

Not sure what to make with that yet but will get there.

**Comment by nicoddemus** [comment]
You can use [pytest-replay](https://github.com/ESSS/pytest-replay) to get a complete list of tests to reproduce the issue in a more controlled manner, without xdist.

**Comment by KelSolaar** [comment]
I think I figured it out thanks to `pytest-replay`!

We can change floating point precision at a global level and we also have a caching system for computationally heavy definitions. We also have a unit test that check that changing the floating point precision works by testing quite a lot of definitions. Doing so, we were caching some values in float32 rather than float64 which was causing the failures. This was happening randomly because of the random-like way `pytest-xdist` distributes the test among workers. We have been lucky in the past in that the test changing precision was probably occurring after any definition that was caching its output.

Closing this one!
